[Warnings] Allow the permissions cog to allow mods to check other user's warnings

Vexed01 opened the pull request:
### Type

- [x] Bugfix
- [x] Enhancement
- [ ] New feature

### Description of the changes
This changes `[p]warnings` to only check other user's warnings and adds `[p]mywarnings` so that the Permissions cog can dictate who can check other's warnings, instead of an unchangeable `is_admin_or_superior` check.
This fixes the duplicate issue #3086.
Closes #2900.

Vexed01 committed:
add le code

Vexed01 committed:
add the real code

i mean it this time

Vexed01 committed:
log le changes

Vexed01 committed:
let's remember how to format .rst

Vexed01 committed:
black formatting

Drapersniper commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -340,33 +340,57 @@ def __init__(self, bot: Red):
+                        mod = discord.utils.get(
+                            self.bot.get_all_members(), id=user_warnings[key]["mod"]
+                        )
+                        if mod is None:
```
Fetch_user should not be used here, since this command is usable but anyone, at this stage we should return a placeholder

mikeshardmind replied:
This is still relevant and should not use fetch_user

Drapersniper commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -340,33 +340,57 @@ def __init__(self, bot: Red):
         await ctx.send(_("User {user} has been warned.").format(user=user))
 
-    @commands.command()
+    @commands.command(usage="")
```
Why is usage set to blank here?

Drapersniper commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -340,33 +340,57 @@ def __init__(self, bot: Red):
-            except AttributeError:
-                userid: int = user
-                user = ctx.guild.get_member(userid)
-                user = user or namedtuple("Member", "id guild")(userid, ctx.guild)
```
instead of doing a if/else check you could shorter is with a simpler “If user is not None” and as such there would be no  no need of the else

Drapersniper commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -340,33 +340,57 @@ def __init__(self, bot: Red):
         if user is None:
             user = ctx.author
-        else:
+        else:  # TODO: remove this else statement in a release after 3.3.0 (will allow users time start using [p]warncheck)
```
Core cogs don’t have to respect the version guarantees and as such breaking changes can be done any time (just a FYI)

Vexed01 replied:
I assume you mean the whole else statement, not just the note?

mikeshardmind reviewed (changes requested):
I'm not going to do a line by line.

Leave the behavior of `warnings` being to check other users, the new command should be checking for your own. I'd name it `mywarnings` but there are plenty of good options for this.

Leaving the behavior of warnings to view others makes sense as this was the more common usage, and the one which the check was trying to enable.

There's also no need to retain compatibility handling, but if you do, the logic is much simpler in this way as you don't need to override usage, just handle a default.

Vexed01 committed:
sinbad's review

use `[p]warnings` for other usuers warnings
user `[p]mywarnings` for users own warnings

Vexed01 commented:
Hmm it didn't dismiss Draper's review as stale... Well I've marked it "resolved" as I've implemented Sinbad's review instead.

Vexed01 committed:
sinbad's review (i think i did it????)

limited python knowledge aint that good... well i'm still learning... gotta start somewhere
also fixes syntax for `[p]warnings`

Vexed01 committed:
lets make the changelog easier to understand

mikeshardmind commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -342,45 +342,57 @@ def __init__(self, bot: Red):
-                return await ctx.send(
-                    warning(_("You are not allowed to check warnings for other users!"))
+    @checks.admin_or_permissions(ban_members=True)
+    async def warnings(self, ctx: commands.Context, user: discord.Member):
```
This should retain the existing ability to check users not in the guild by accepting an id still.

mikeshardmind commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -342,45 +342,57 @@ def __init__(self, bot: Red):
-            if not await is_admin_or_superior(self.bot, ctx.author):
-                return await ctx.send(
-                    warning(_("You are not allowed to check warnings for other users!"))
+    @checks.admin_or_permissions(ban_members=True)
```
This should just be `checks.admin()` matching the prior behavior handled inside the command.

mikeshardmind commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -342,45 +342,57 @@ def __init__(self, bot: Red):
+                    mod = ctx.guild.get_member(user_warnings[key]["mod"])
+                    if mod is None:
+                        mod = discord.utils.get(
+                            self.bot.get_all_members(), id=user_warnings[key]["mod"]
```
use `ctx.guild.get_member`

mikeshardmind reviewed (changes requested):
There's more changes needed on this.

Vexed01 committed:
sinbad's review... god it's like he's coding the whole thing for me

- re-allow checking warnings of users not in guild with ids
- change the check
- use ctx.guild.get_member

Vexed01 committed:
oops... lets put the changelog in the correct folder

mikeshardmind commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -342,45 +342,57 @@ def __init__(self, bot: Red):
+                    mod = ctx.guild.get_member(user_warnings[key]["mod"])
+                    if mod is None:
+                        mod = discord.utils.get(
+                            ctx.guild.get_member(), id=user_warnings[key]["mod"]
```
Not really what I meant by that...

My point wasn't to try and make `guild.get_member` work as part of utils.get (this is already properly handled above), but that while this was being changed, we should avoid leaking info from other guilds and not attempt to match outside the current guild.

Vexed01 commented:
See https://github.com/Cog-Creators/Red-DiscordBot/pull/3091#issuecomment-570160913.
